source/common/stats/allocator_impl.cc

```diff
@@ -226,12 +223,42 @@ class GaugeImpl : public StatsSharedImpl<Gauge> {
       break;
     }
   }
+
+private:
+  std::atomic<uint64_t> value_{0};
+};
+
+class TextReadoutImpl : public StatsSharedImpl<TextReadout> {
+public:
+  TextReadoutImpl(StatName name, AllocatorImpl& alloc, StatName tag_extracted_name,
+                  const StatNameTagVector& stat_name_tags)
+      : StatsSharedImpl(name, alloc, tag_extracted_name, stat_name_tags) {}
+
+  void removeFromSetLockHeld() EXCLUSIVE_LOCKS_REQUIRED(alloc_.mutex_) override {
+    const size_t count = alloc_.text_readouts_.erase(statName());
+    ASSERT(count == 1);
+  }
+
+  // Stats::TextReadout
+  void set(const std::string& value) override {
+    absl::MutexLock lock(&mutex_);
```

Comment:
Great suggestion! Done.

Reply:
actually I just thought of another idea; may be this would be more intuitive?

Leave this API as is, but change the implementation to copy the input param to a local without taking the lock, and then std::move it into the structure after taking the lock.

Reply:
I agree that having `set()` method move input value is quite surprising, and I was going to suggest to keep `const std::string&` and add `swap(std::string& value)` method, but I did find quite a few methods that take `&&`, so I've scraped that. :)

To be honest I would argue that lock contention is not an issue as I don't expect to have text readout values changed racily from multiple threads. 
I might be coming from a very narrow use case point of view, but in that use case the value of the text readout is unlikely to change ever.
What use case are you thinking about?